File: Services/pkg/user/cache.go
package user

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"
	"github.com/gofiber/storage/redis"
)

const (
	cacheTTL     = 15 * time.Minute
	cacheTTLList = 1 * time.Minute
)

type userCache struct {
	store *redis.Storage
}

func newUserCache(store *redis.Storage) *userCache {
	return &userCache{store: store}
}

func (c *userCache) Get(id uint) (*models.User, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%d", models.PkgUserGetUser, id)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var user models.User
	if err := json.Unmarshal(bytes, &user); err != nil {
		return nil, false
	}
	return &user, true
}

func (c *userCache) Set(user *models.User) {
	if c.store == nil || user == nil {
		return
	}

	go func(u models.User) {
		key := fmt.Sprintf("%s:%d", models.PkgUserGetUser, u.ID)
		if bytes, err := json.Marshal(u); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTL))
		}
	}(*user)
}

func (c *userCache) Invalidate(id uint, email string) {
	if c.store == nil {
		return
	}
	go func() {
		key := fmt.Sprintf("%s:%d", models.PkgUserGetUser, id)
		_ = c.store.Delete(key)
	}()
	go func() {
		key := fmt.Sprintf("%s:%s", models.PkgUserGetUserGmail, email)
		_ = c.store.Delete(key)
	}()
	c.InvalidateAllLists()
}

func (c *userCache) GetByEmail(email string) (*models.User, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%s", models.PkgUserGetUserGmail, email)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var user models.User
	if err := json.Unmarshal(bytes, &user); err != nil {
		return nil, false
	}
	return &user, true
}

func (c *userCache) SetByEmail(email string, user *models.User) {
	if c.store == nil || user == nil {
		return
	}
	go func(u models.User) {
		key := fmt.Sprintf("%s:%s", models.PkgUserGetUserGmail, email)
		if bytes, err := json.Marshal(u); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTL))
		}
	}(*user)
}
func (c *userCache) InvalidateAllLists() {
	if c.store == nil {
		return
	}
	pattern := fmt.Sprintf("%s:*", models.PkgUserGetUsers)
	conn := c.store.Conn()
	ctx := context.Background()
	keys, err := conn.Keys(ctx, pattern).Result()
	if err != nil {
		return
	}
	if len(keys) == 0 {
		return
	}
	const numWorkers = 100
	keyChan := make(chan string, len(keys))
	var wg sync.WaitGroup
	workerCount := numWorkers
	if len(keys) < numWorkers {
		workerCount = len(keys)
	}
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for key := range keyChan {
				_ = c.store.Delete(key)
			}
		}()
	}
	for _, key := range keys {
		keyChan <- key
	}
	close(keyChan)
	wg.Wait()
}
func (c *userCache) GetList(pagination models.Pagination, search models.Search) ([]models.User, *models.Pagination, *models.Search, bool) {
	if c.store == nil {
		return nil, nil, nil, false
	}
	key := fmt.Sprintf("%s:%s:%s", models.PkgUserGetUsers, pagination.GetPaginationString(), search.GetSearchString())
	fmt.Println("key get", key)
	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, nil, nil, false
	}

	var usersCache UsersCache
	if err := json.Unmarshal(bytes, &usersCache); err != nil {
		fmt.Println("err", err)
		return nil, nil, nil, false
	}
	return usersCache.Users, &usersCache.Pagination, &usersCache.Search, true
}

func (c *userCache) SetList(pagination models.Pagination, search models.Search, users []models.User) {
	if c.store == nil {
		return
	}
	go func(us []models.User, p models.Pagination, s models.Search) {
		usersCache := UsersCache{
			Users:      us,
			Pagination: p,
			Search:     s,
		}
		key := fmt.Sprintf("%s:%s:%s", models.PkgUserGetUsers, p.GetPaginationString(), s.GetSearchString())
		if bytes, err := json.Marshal(usersCache); err == nil {
			fmt.Println("key set", key)
			err = c.store.Set(key, bytes, utils.RandomJitter(cacheTTLList))
			if err != nil {
				fmt.Println(err)
			}
		}
	}(users, pagination, search)
}

type UsersCache struct {
	Users []models.User `json:"users"`
	models.Pagination
	models.Search
}

File: hono-clean-architecture/pkg/user/cache.ts
import Redis from "ioredis";
import { UserResponse, Pagination, Search, CacheKeys } from "../models";
import { randomJitter } from "../utils";

const CACHE_TTL = 15 * 60 * 1000;
const CACHE_TTL_LIST = 1 * 60 * 1000;

export class UserCache {
  private store: Redis | null;

  constructor(store: Redis | null) {
    this.store = store;
  }

  async get(id: number): Promise<UserResponse | null> {
    if (!this.store) return null;

    const key = `${CacheKeys.PkgUserGetUser}:${id}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as UserResponse;
    } catch {
      return null;
    }
  }

  async set(user: UserResponse): Promise<void> {
    if (!this.store || !user) return;

    const key = `${CacheKeys.PkgUserGetUser}:${user.id}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(user), "PX", ttl);
  }

  async getByEmail(email: string): Promise<UserResponse | null> {
    if (!this.store) return null;

    const key = `${CacheKeys.PkgUserGetUserByEmail}:${email}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as UserResponse;
    } catch {
      return null;
    }
  }

  async setByEmail(email: string, user: UserResponse): Promise<void> {
    if (!this.store || !user) return;

    const key = `${CacheKeys.PkgUserGetUserByEmail}:${email}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(user), "PX", ttl);
  }

  async getList(
    pagination: Pagination,
    search: Search
  ): Promise<{ users: UserResponse[]; pagination: Pagination; search: Search } | null> {
    if (!this.store) return null;

    const cacheKey = this.buildListCacheKey(pagination, search);
    const data = await this.store.get(cacheKey);

    if (!data) return null;

    try {
      return JSON.parse(data);
    } catch {
      return null;
    }
  }

  async setList(users: UserResponse[], pagination: Pagination, search: Search): Promise<void> {
    if (!this.store) return;

    const cacheKey = this.buildListCacheKey(pagination, search);
    const ttl = randomJitter(CACHE_TTL_LIST);

    await this.store.set(cacheKey, JSON.stringify({ users, pagination, search }), "PX", ttl);
  }

  async invalidate(id: number, email?: string): Promise<void> {
    if (!this.store) return;

    const keys: string[] = [`${CacheKeys.PkgUserGetUser}:${id}`];
    if (email) {
      keys.push(`${CacheKeys.PkgUserGetUserByEmail}:${email}`);
    }

    await this.store.del(...keys);
  }

  async invalidateAllLists(): Promise<void> {
    if (!this.store) return;

    const pattern = `${CacheKeys.PkgUserGetList}:*`;
    const keys = await this.store.keys(pattern);

    if (keys.length > 0) {
      await this.store.del(...keys);
    }
  }

  private buildListCacheKey(pagination: Pagination, search: Search): string {
    const paginationStr = `p${pagination.page}_pp${pagination.perPage}_ob${pagination.orderBy || ""}_o${pagination.order || ""}`;
    const searchStr = `k${search.keyword || ""}_c${search.column || ""}`;
    return `${CacheKeys.PkgUserGetList}:${paginationStr}:${searchStr}`;
  }
}

export function newUserCache(store: Redis | null): UserCache {
  return new UserCache(store);
}
